gossip/gossip_test.go

```diff
@@ -0,0 +1,149 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package gossip
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/ava-labs/avalanchego/codec"
+	"github.com/ava-labs/avalanchego/codec/linearcodec"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/network/p2p"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/utils/set"
+	"github.com/ava-labs/avalanchego/utils/units"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/require"
+)
+
+func TestGossiperShutdown(t *testing.T) {
+	config := Config{Frequency: time.Second}
+	gossiper := NewGossiper[testTx](config, nil, nil, nil, 0)
+	done := make(chan struct{})
+	wg := &sync.WaitGroup{}
+
+	wg.Add(1)
+	go gossiper.Gossip(done, wg)
+
+	close(done)
+	wg.Wait()
+}
+
+func TestGossiperGossip(t *testing.T) {
+	tests := []struct {
+		name      string
+		requester []*testTx // what we have
+		responder []*testTx // what the peer we're requesting gossip from has
+		expected  []*testTx // what we should have after a gossip cycle
+	}{
+		{
+			name: "no gossip - no one knows anything",
+		},
+		{
+			name:      "no gossip - requester knows more than responder",
+			requester: []*testTx{{hash: Hash{0}}},
+			expected:  []*testTx{{hash: Hash{0}}},
+		},
+		{
+			name:      "no gossip - requester knows everything responder knows",
+			requester: []*testTx{{hash: Hash{0}}},
+			responder: []*testTx{{hash: Hash{0}}},
+			expected:  []*testTx{{hash: Hash{0}}},
+		},
+		{
+			name:      "gossip - requester knows nothing",
+			responder: []*testTx{{hash: Hash{0}}},
+			expected:  []*testTx{{hash: Hash{0}}},
+		},
+		{
+			name:      "gossip - requester knows less than responder",
+			requester: []*testTx{{hash: Hash{0}}},
+			responder: []*testTx{{hash: Hash{0}}, {hash: Hash{1}}},
+			expected:  []*testTx{{hash: Hash{0}}, {hash: Hash{1}}},
+		},
```

Comment:
Can we add a check within the test to ensure that the transactions returned by the responder are only the transactions that should be included? Ie in this case, we check to make sure that the requester has 0 and 1 after receiving the gossip response, but if the response included 0 even when it should not, this test would still pass.

Reply:
Added a check for this